CtD Goals Step: Add new goals and rework intents in the existing signup goals step

p-jackson opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Fixes https://github.com/Automattic/wp-calypso/issues/95640. Fixes https://github.com/Automattic/wp-calypso/issues/95644.

## Proposed Changes

Updates the goals screen following the new design: p9Jlb4-emx-p2
This is only the first step in a broader update to the signup flow, so this change needs to work nicely with what's coming and with how things currently work. Therefore:

- The `Skip to dashboard` button sends the user to Calypso home and not `wp-admin` yet: this will be done in a separate PR.
- Choosing goals still sets an implicit `intent` under the hood, since that's how so much of the flow navigation works
- Picking any of the items below the goals selector record a different intent: `import`, `difm` and `build`, respectively

**Large Screen**

![image](https://github.com/user-attachments/assets/57ba2380-fd8c-44eb-96a2-0c1233c73450)

**Small Screen**
| Above the fold | Below the fold |
| --- | --- |
| ![calypso localhost_3000_setup_site-setup_goals_siteSlug=zaguiini453 wordpress com(iPhone XR)](https://github.com/user-attachments/assets/56a71eed-03f2-4e1b-8e28-1dfbe215a974) | ![calypso localhost_3000_setup_site-setup_goals_siteSlug=zaguiini453 wordpress com(iPhone XR) (1)](https://github.com/user-attachments/assets/e9936eec-c675-44c7-9865-8a84cbe70455) |

## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

These changes are part of a larger effort to to focus the signup and onboard process more on goals and less on bespoke flows. Read more: p58i-irT-p2

## Testing Instructions

### Goals

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

You can either test the goals screen by creating new sites (starting from `calypso.localhost:3000/start`) or by using the goals screen with an existing site. **Note that using the `site-setup` flow existing sites will override some of their data**.

Test the `site-setup` flow with an existing site by visiting `calypso.localhost:3000/setup/site-setup/goals?siteSlug={{ YOUR SITE HERE }}`

After you complete the flow the goals will be saved to your site. You can check the goals currently saved to your site using [the dev console](https://developer.wordpress.com/docs/api/console/). Use `GET` `WP REST API` `wpcom/v2` `/sites/{{ YOUR SITE HERE}}/site-goals`:

![CleanShot 2024-10-25 at 20 27 22@2x](https://github.com/user-attachments/assets/a784e092-adac-4c1c-b9b7-f6a89c2104a7)

Things to check:
* Confirm the `site-setup` flow is the only flow using the `goals` step. Otherwise we would need to make sure we weren't breaking anything with this flow
* Test mobile
* Confirm tracks events still work as expected. Any funnel or dashboard that was watching these metrics shouldn't break due to the change in goals.
* Goals are correctly saved correctly after completing the flow
* Selecting `Publish a blog` or `Sell ...` and clicking **Next** takes you to a step related to that goal
	* Otherwise you choose a design, similar to if you choose the `Other` goal in the current goals screen.
* The custom DIFM and import flows don't save goals.

## Tracks events

### Individual goals recording

A new event called `calypso_signup_goals_single_select` is dispatched N times and with `goal: [goal]` property according to the selected goals

![image](https://github.com/user-attachments/assets/e694e7ad-3ef6-42fe-840d-f0ab2dfbd4d0)

### Skip

#### Clicking "Skip" records no goals and stores the `build` intent

![image](https://github.com/user-attachments/assets/d78a9da0-c31a-4c94-9924-01f73bfea966)

#### The submit action event includes the `action: skip` property

![image](https://github.com/user-attachments/assets/7e509825-77d2-4e51-ae1c-9205b8a5b028)

### Next

#### Selecting store goals (sell X) and clicking "Next" stores the appropriate goals and intents

![image](https://github.com/user-attachments/assets/c91f6a5e-f0dd-44f9-9a80-85f5365f18ec)

#### The submit action event includes the `action: next` property

![image](https://github.com/user-attachments/assets/39f955b3-4ba4-4de3-bccf-a7f5de9628cb)

### Import

#### Selecting "Import or migrate an existing site" does not store goals and records the `import` intent

![image](https://github.com/user-attachments/assets/47ddaad7-9871-43f1-abb6-95bb634c571c)

#### The submit action event includes the `action: import` property

![image](https://github.com/user-attachments/assets/eb42ef04-4fa8-4590-9eaa-7b347bbfa0be)

### DIFM

#### Selecting "Let us build a custom site for you" does not store goals and records the `difm` intent

![image](https://github.com/user-attachments/assets/d230ccbf-4353-40d3-8c72-a20ae46b79d9)

#### The submit action event includes the `action: difm` property

![image](https://github.com/user-attachments/assets/b4d5c683-4d01-409c-8748-e47ace950962)

### Skip to dashboard

#### Selecting "Skip to dashboard" does not store goals and records the `build` intent

![image](https://github.com/user-attachments/assets/e97b969e-e485-4cd6-a58f-fbcc76b47161)

#### The submit action event includes the `action: dashboard` property

![image](https://github.com/user-attachments/assets/7baebaa6-bd06-4b52-bda9-dfe373b67cc2)

p-jackson committed:
Clean up deprecated code

p-jackson committed:
Reword goal step's header and sub header copy

p-jackson committed:
Reword/add new goals

p-jackson committed:
Add import and DIFM nav links to the bottom of the screen

p-jackson committed:
Update step navigation controls

The "Next step" button is now in the header. The "Skip to dashboard"
button is at the bottom of the page.

Although the mockup calls the skip button "Go to WordPress Admin", we
aren't yet ready to default all new sites to wp-admin, so I'll keep
calling it "Dashboard" for now.

p-jackson committed:
Ensure new sell goals select the sell flow (for now)

The sell flow will go away, however for now we want to ensure selecting
a sell goals sets the sell intent, so that proceeding in the flow goes
to the sell step.

I was able to tidy the code. There might have been a point in time when
the GOAL_TO_INTENT_MAP was useful, however now all it was doing was
setting the write intent.

p-jackson committed:
Style bottom links using CSS grid so things wrap nicely on mobile

p-jackson committed:
Update header padding/wrapping on mobile and desktop

zaguiini committed:
Update Offer a contact form copy to Create a contact form

zaguiini committed:
Randomize items on page load

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/16949432

Some locales (Hebrew, Japanese) have been temporarily machine-translated due to translator availability. All other translations are usually ready within a few days. Untranslated and machine-translated strings will be sent for translation next Monday and are expected to be completed by the following Friday.

Thank you @p-jackson for including a screenshot in the description! This is really helpful for our translators.

zaguiini committed:
Update actions according to the designs

zaguiini commented on packages/viewport/src/index.ts:
```diff
@@ -95,6 +96,7 @@ function createMediaQueryList( args?: QueryOption ): QueryItem {
 // See https://github.com/WordPress/gutenberg/blob/f1d0bd550f85f5fa4279a3fdb9a2b9c28a7544c6/packages/base-styles/_mixins.scss#L27
 const mediaQueryOptions: Record< string, QueryOption > = {
 	'<480px': { max: 480 },
+	'<600px': { max: 600 },
```
@p-jackson, we probably don't want to add another breakpoint here, but this one is the actual value of `break-small` coming from Gutenberg, and I believe that's what we want moving forward. (That's why I see `>660px` being used with a "deprecated" symbol within CSS.)

Should we keep this here? Is there a better way to reference this specific breakpoint within React?

p-jackson replied:
If we're going to move the **Continue** button from the main body into the toolbar I can't think of a way to do this that doesn't involve JavaScript unfortunately.

I think `useViewportMatch` might be better though, since it's already a Core thing and uses the standard GB break points.
[Like this example](https://github.com/Automattic/wp-calypso/blob/a5d5804c343c648deb0dcc473142e82db4862175/client/signup/steps/courses/index.tsx#L22)

Also what I like better is it consistently enforces either `<` or `>=`, because of course `<600` is not the exact opposite of `>600`. That confusion I think is creating this behaviour.


https://github.com/user-attachments/assets/97045a9a-1536-49ab-8925-d7b2a723ad79

It's probably not that big a deal. And switching to `useViewportMatch` might not fix it immediately. The stepper frameworks mobile toolbar might not be using the standard mixing breakpoint, it might be using the `break-small` variable direction, which means the logic could be a bit busted still.

zaguiini committed:
Fix Skip to Dashboard click behavior

autumnfjeld commented:
@zaguiini 

### 1) The Next button is out of alignment:
<img width="1840" alt="image" src="https://github.com/user-attachments/assets/467faef2-72da-4925-a832-b91dccdddd85">


### 2)
@nuriapenya  What do you think we should do about randomization for when a user goes forward and then back and the goals are rearranged?  I'm not sure if we should address this or if it is a big deal.  
I  will check metrics on going back from the Design Picker (DP), which gives a ballpark estimate of how many people are doing this:
* Go from Goals screen to DP
* Click back
* See Goals screen again

autumnfjeld commented:
@nuriapenya @zaguiini 
Looks like the design decisions are still being discussed: p1730154921739109/1730065270.612889-slack-C07H21B2W59

zaguiini committed:
Preserve randomized goals upon refresh

zaguiini commented on client/landing/stepper/utils/shuffleArray.ts:
```diff
@@ -0,0 +1,36 @@
+
+let localSeed: string | null = null;
+
+const getSeed = () => {
```
@p-jackson, here's our seed function. If that fails, it tries fetching/saving to session storage and falls back to a local variable. Let me know what you think.

p-jackson replied:
Yes this looks like the right approach to me 👍 
The local variable will be faster so you could switch the logic to `localSeed ?? sessionStorage.getItem( ... )`? Use the local variable like a cache?

I wasn't expecting the randomisation function to look like `Math.sin( ++seed ) * 10000` 😅 but TIL that native JavaScript doesn't support the ability to provide your own seed! I'd suggested it just assuming it was a built-in like every other language 🤷

ChatGPT suggested a function like this, which would provide less regular results than `sin()`
```
function seededRandom(seed) {
  let value = seed;
  return function() {
    value = (value * 9301 + 49297) % 233280;
    return value / 233280;
  };
}
```

zaguiini committed:
Allow skipping to next step as well as skipping to dashboard

zaguiini committed:
Improve viewport breakpoint handling

zaguiini committed:
Improve goal-shuffling logic

zaguiini committed:
Fix style inconsistencies in production build and consolidate selectors

zaguiini committed:
Record intent as a Tracks event whenever clicking Skip to dashboard

zaguiini committed:
Rewrite new goal values with an underscore so it conforms to Track property names expectations

zaguiini committed:
Stop tracking the import property in the calypso_signup_intent_select event

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/16949432).

p-jackson commented on client/landing/stepper/declarative-flow/internals/steps-repository/goals/index.tsx:
```diff
@@ -15,14 +15,18 @@ import type { Step } from '../../types';
 import type { OnboardSelect } from '@automattic/data-stores';
 import './style.scss';
 
+type KebabToSnakeCase< S extends string > = S extends `${ infer T }${ infer U }`
```
Nice 🤣

p-jackson commented on client/landing/stepper/declarative-flow/internals/steps-repository/goals/index.tsx:
```diff
@@ -81,35 +84,41 @@ const GoalsStep: Step = ( { navigation } ) => {
+	const recordIntentSelectTracksEvent = ( intent: Onboard.SiteIntent ) => {
 		const eventProperties = {
 			intent,
-			import: hasImportGoal,
```
Removing properties can break existing funnels. I did a non-exhaustive search of the existing funnels which use `calypso_signup_intent_select` and didn't find anything using the `import` prop (specifically filtering for funnels created by those involved in [the original PR](https://github.com/Automattic/wp-calypso/pull/64725)), so maybe this is safe to do? But if there's no pressing reason to drop the prop, maybe we should keep it?

zaguiini replied:
I'm not sure keeping it helps much 🤷🏻  It would always be `false` since we were considering passing an empty array (and completely removed the parameter as of my last commit).  

Anyway, I went through all funnels containing the event, and the `import` property specifically filters none of them, so it's safe to ignore it. We have created only one funnel this year containing the event, and the last-but-one was calculated 12 months ago. It's a pretty safe ignore, IMHO.

p-jackson replied:
Oh yeah, always false, you're right 😅

p-jackson reviewed (commented):
I can't ✅ approve the PR I authored, but I approve this PR :)

nuriapenya commented:
Looks great @zaguiini 

For some reason, the checkbox on mobile pops into a bigger size. It should just stay the same:
![screen 2024-11-04 at 14 08 03](https://github.com/user-attachments/assets/de9de868-313a-44c9-85b6-651c27e36380)

The outline on the "Skip" link still shows the old blue. Could we update that? 🙏🏽 Small one but hopefully an easy fix.
![screen 2024-11-04 at 14 11 55](https://github.com/user-attachments/assets/7fdc1f7c-681b-4120-b19d-ba3837f316ec)

Could we also decrease the gap here to 24px (instead of 32px?)
![screen 2024-11-04 at 14 23 07](https://github.com/user-attachments/assets/7f9c44b7-9606-437a-9604-47f7c5602978)

And finally, two more design tweaks (this is an amend on Figma):
* Could we make the CTA 40px high on desktop?
* Could we add an outline border to the hover? 

Figma → [Here](https://www.figma.com/design/W9xI27S6Swvw5Ku21EbZvn/Dotcom-Onboarding?node-id=5824-24222&t=lwcxX2qrZN4pp4hf-4)

![screen 2024-11-04 at 14 48 22](https://github.com/user-attachments/assets/7fb9d2c9-7f2d-4ba1-97fe-a4f0ccf306e4)

I am working on an standardization of our onboarding styles, and I hope to put something on P2 soon. But for now, the changes on the Goals screen can set the foundation of it.

zaguiini committed:
Make Next button 40px in height

zaguiini committed:
Change content gap to 24px

zaguiini committed:
Match hover and active styles in the SelectCardCheckbox component

zaguiini committed:
Skip button should match new accent color

p-jackson committed:
New shuffle array module uses kebab case

p-jackson committed:
Fix SelectGoal tests to be robust in the face of randomness

There was a 1% chance the "randomizes goals between sessions" test would
fail because `shuffleArray` randomly used the same order.

lsl commented on client/landing/stepper/declarative-flow/site-setup-flow.ts:
```diff
@@ -627,13 +632,6 @@ const siteSetupFlow: Flow = {
-					return exitFlow( `/home/${ siteId ?? siteSlug }`, {
-						skipLaunchpad: true,
-					} );
-
```
What's the goToNext => submit swap doing here? 

Was it intentional to no longer setIntent to the main build flow?

p-jackson replied:
`goNext` is marked as deprecated, and we use `submit` instead.

Despite being deprecated, the buttons in the step header component still only support using `goNext`, and so the "skip to dashboard" button in the header used to use `goNext`. In fact you can see that the code assumes that if `goNext` is called at all during the goals step then it must be a "skip to dashboard" action.

However now the "skip to dashboard" button has moved to the main body of the step and we can use `submit` navigation method instead.

The "skip" button in the header now "skips to the next step in the flow", not "skips to the dashboard"

lsl replied:
Ah, that makes sense, thanks.

> setIntent( SiteIntent.Build );

It wasn't clear, is that still being triggered somewhere for skips?

p-jackson replied:
Yes skipping will set `build`, and leaving the flow via DIFM or the Import link will set the expected intent.

zaguiini committed:
Disambiguate event submissions by using the action property

zaguiini committed:
Dispatch one goals_single_select Tracks event for each selected goal

p-jackson committed:
Introduces A/B test for whether to show the new goals

p-jackson committed:
Big Sky eligibility supports new goal types

See peGwbA-2BZ-p2#comment-3978. The Big Sky experiment eligibility has
gone from opt-in based on some goals, to opt-out of goals which Big Sky
doesn't do will with yet.

p-jackson committed:
Reinstate old copy for the experiment's control variation

p-jackson committed:
Reinstate old navigation for experiment's control variation

p-jackson committed:
Unit tests use new `isAddedGoalsExp` prop

p-jackson committed:
Remove is_added_goals_exp prop

This particular prop won't be useful to any hypothetical Big Sky
experiment

p-jackson committed:
Skip track is only for top-right, submit is only for main CTA

For better compatibility between the control variant and what's already
in production, the `calypso_signup_actions_submit_step` action is only
sent by the main CTA, `calypso_signup_skip_step` is only sent by the
top-left skip button. Thus the `action` prop has been removed from
`calypso_signup_actions_submit_step` events.
So we can distinguish nav clicks in the treatment variant, a new
`calypso_signup_goals_nav_click` event has been added, which includes
the `action` prop.
